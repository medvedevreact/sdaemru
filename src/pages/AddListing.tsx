import { useEffect, useRef, useState } from "react";
import { AddListingForms } from "../components/AddListingForm/AddListingForms";

import { AppartmentItem } from "../types";
import { AutoItem } from "../types";
import { HouseItem } from "../types";
import { addListing } from "../store/listingsSlice";
import { useAppDispatch } from "../store";
import CustomButton from "../ui/Button/CustomButton";

type FormFields = {
  appartments: AppartmentItem;
  houses: HouseItem;
  auto: AutoItem;
};
const formFields: FormFields = {
  appartments: {
    title: "",
    description: "",
    price_per_day: 0,
    location: "",
    metro: "",
    district: "",
    rooms: "",
    photo: ["", "", "", ""],
    owner: {
      name: "",
      phone: "",
      email: "",
    },
  },
  houses: {
    title: "",
    description: "",
    price_per_day: 0,
    location: "",
    metro: "",
    district: "",
    size: "",
    photo: ["", "", "", ""],
    owner: {
      name: "",
      phone: "",
      email: "",
    },
  },
  auto: {
    title: "",
    description: "",
    price_per_day: 0,
    location: "",
    metro: "",
    district: "",
    type: "",
    photo: ["", "", "", ""],
    owner: {
      name: "",
      phone: "",
      email: "",
    },
  },
};

const convertNamesObj = {
  appartments: "Квартиры",
  houses: "Коттеджи",
  auto: "Авто",
};

const convertOptions = {
  appartments: "rooms",
  houses: "size",
  auto: "type",
};

export const AddListing = () => {
  const [activeCategory, setActiveCategory] = useState("Квартиры");
  const dispatch = useAppDispatch();
  const fileInputRef = useRef<HTMLInputElement | null>(null);
  const resetFileInputs = () => {
    if (fileInputRef.current) {
      fileInputRef.current.value = "";
    }
    if (secondPhotoRef.current) {
      secondPhotoRef.current.value = "";
    }
    if (thirdPhotoRef.current) {
      thirdPhotoRef.current.value = "";
    }
    if (fourthPhotoRef.current) {
      fourthPhotoRef.current.value = "";
    }
  };
  const secondPhotoRef = useRef<HTMLInputElement | null>(null);
  const thirdPhotoRef = useRef<HTMLInputElement | null>(null);
  const fourthPhotoRef = useRef<HTMLInputElement | null>(null);

  const [showMessage, setShowMessage] = useState(false);

  const activeCategoryKey = Object.keys(convertNamesObj).find(
    (key) =>
      convertNamesObj[key as keyof typeof convertNamesObj] === activeCategory
  ) as keyof typeof formFields;
  const [listingObject, setListingObject] = useState(
    formFields[activeCategoryKey]
  );
  const option = convertOptions[activeCategoryKey];

  const allPhotos = listingObject.photo.every((el) => el !== "");

  const handleClick = () => {
    if (
      listingObject.location !== "" &&
      listingObject.description !== "" &&
      allPhotos &&
      listingObject.price_per_day !== 0 &&
      option &&
      listingObject[option as keyof typeof listingObject] !== "" &&
      listingObject.metro !== "" &&
      listingObject.title !== ""
    ) {
      dispatch(
        addListing({
          category: activeCategoryKey as string,
          listing: listingObject,
        })
      );
      setListingObject(formFields[activeCategoryKey]);
      resetFileInputs();
    } else {
      setShowMessage(true);
    }
  };

  useEffect(() => {
    let timer: ReturnType<typeof setTimeout>;
    if (showMessage) {
      timer = setTimeout(() => {
        setShowMessage(false);
      }, 3000);
    }
    return () => clearTimeout(timer);
  }, [showMessage]);

  return (
    <div className="container">
      <div className="addListingWrapper">
        <ul className="addListingTitle">
          {Object.entries(convertNamesObj).map(([key, value]) => (
            <li
              key={key}
              onClick={() => {
                setActiveCategory(value);
                setListingObject(formFields[activeCategoryKey]);
              }}
              className={`addListingTitleItem ${
                activeCategory === value ? "active" : ""
              }`}
            >
              {value}
            </li>
          ))}
        </ul>
        <div className="addListingContent">
          <div>
            <h2 className="addListingContentTitle">
              Добавление объявления {activeCategory}
            </h2>
            <p className="addListingContentText">
              1. Введите адрес объекта, который вы собираетесь сдавать в аренду.
              <br></br>
              2. Укажите название и описание.
              <br></br>
              3. Добавьте фотографии.
              <br></br>
              4. Укажите цену.
              <br></br>
              5. Укажите дополнительные характеристики объекта в указанных
              полях.
            </p>
          </div>
          <div>
            <AddListingForms
              fileInputRef={fileInputRef}
              activeCategoryKey={activeCategoryKey}
              listingObject={listingObject}
              setListingObject={setListingObject}
              secondPhotoRef={secondPhotoRef}
              thirdPhotoRef={thirdPhotoRef}
              fourthPhotoRef={fourthPhotoRef}
            />

            <CustomButton onClick={handleClick} size={"medium"}>
              Добавить объявление
            </CustomButton>
            {showMessage && (
              <p className="addListingNotCompleted">
                Заполните все необходимые поля
              </p>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};
