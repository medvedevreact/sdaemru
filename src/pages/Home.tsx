import { Search } from "../components/Search/Search";
import { Gallery } from "../components/Gallery/Gallery";

export const Home = () => {
  return (
    <div className="container">
      <div className="home-page">
        <Search />
        <Gallery />
      </div>
    </div>
  );
};
