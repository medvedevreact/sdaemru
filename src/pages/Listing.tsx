import React, { useEffect, useState } from "react";
import { ListingUpper } from "../components/ListingUpper/ListingUpper";
import { useParams } from "react-router-dom";
import axios from "axios";
import { ListingMiddle } from "../components/ListingMiddle/ListingMiddle";

export const Listing = () => {
  const { id } = useParams();
  const { category } = useParams();

  const [listing, setListing] = useState(null);
  useEffect(() => {
    axios.get(`http://localhost:3000/${category}/${id}`).then((resp) => {
      setListing(resp.data);
    });
  }, [id, category]);
  return (
    <div className="pageWrapper">
      <div className="container">
        {listing && (
          <>
            <ListingUpper listing={listing} />
            <ListingMiddle listing={listing} />
          </>
        )}
      </div>
    </div>
  );
};
