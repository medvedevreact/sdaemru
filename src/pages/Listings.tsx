import { useEffect } from "react";
import { ListingsTitle } from "../components/ListingsTitle/ListingsTitle";
import { useParams } from "react-router-dom";
import { clearFilter, fetchListings } from "../store/listingsSlice";
import { ListingsItems } from "../components/ListingsItems/ListingsItems";
import { useUnmount } from "usehooks-ts";
import { useAppDispatch, useAppSelector } from "../store";

export const Listings = () => {
  const { category } = useParams<{ category: string }>();
  const safeCategory = category ?? "defaultCategory";

  const listings = useAppSelector((state) => state.listings.listings);
  const filterOptions = useAppSelector((state) => state.listings.filter);
  const fromPrice = useAppSelector((state) => state.listings.price.fromPrice);
  const toPrice = useAppSelector((state) => state.listings.price.toPrice);

  const dispatch = useAppDispatch();

  useEffect(() => {
    const getListings = () => {
      dispatch(
        fetchListings({
          category,
          filter: filterOptions,
          fromPrice: fromPrice,
          toPrice: toPrice,
        })
      );
    };
    getListings();
  }, [filterOptions, category, dispatch, fromPrice, toPrice]);

  useUnmount(() => {
    dispatch(clearFilter());
  });

  return (
    <div>
      <ListingsTitle category={safeCategory} />
      <div className="pageWrapper">
        {listings.length > 0 && <ListingsItems items={listings} />}
      </div>
    </div>
  );
};
