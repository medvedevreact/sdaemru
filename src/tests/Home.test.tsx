import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { Home } from "../pages/Home";

jest.mock("../components/Search/Search", () => ({
  Search: () => <div>Search Component</div>,
}));

jest.mock("../components/Gallery/Gallery", () => ({
  Gallery: () => <div>Gallery Component</div>,
}));

describe("Home component", () => {
  it("renders correctly and matches the snapshot", () => {
    const { asFragment } = render(<Home />);
    expect(asFragment()).toMatchSnapshot();
  });
});
