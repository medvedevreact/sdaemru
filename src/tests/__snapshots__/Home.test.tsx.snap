// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component renders correctly and matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="home-page"
    >
      <div>
        Search Component
      </div>
      <div>
        Gallery Component
      </div>
    </div>
  </div>
</DocumentFragment>
`;
